### DISCUSSION TITLE
Introduce `kimm`: a collection of image models, blocks and layers written in Keras3

### DISCUSSION THREAD

**Discussion by james77777778** [keras-team/keras#19718] [Show and tell]
Repository: https://github.com/james77777778/keras-image-models

If you want to try a different model that isn't available in Keras, you can explore this library.
Feel free to contact me if you encounter any issues or have ideas.

## Introduction

**K**eras **Im**age **M**odels (`kimm`) aims to provide SOTA models with pretrained weights in a user-friendly manner.

Features:

- 🚀 A model zoo where almost all models come with **pre-trained weights on ImageNet**.
- 🧰 Providing APIs to export models to `.tflite` and `.onnx`.
- 🔧 Supporting the **reparameterization** technique.
- ✨ Integrated with **feature extraction** capability.

## Usage

```python
import keras
import kimm
import numpy as np


# List available models
print(kimm.list_models("mobileone", weights="imagenet"))  # ['MobileOneS0', ...]

# Initialize model with pretrained ImageNet weights
x = keras.random.uniform([1, 224, 224, 3])
model = kimm.models.MobileOneS0()
y = model.predict(x)
print(y.shape)  # (1, 1000)

# Get reparameterized model by kimm.utils.get_reparameterized_model
reparameterized_model = kimm.utils.get_reparameterized_model(model)
y2 = reparameterized_model.predict(x)
np.testing.assert_allclose(
    keras.ops.convert_to_numpy(y), keras.ops.convert_to_numpy(y2), atol=1e-5
)

# Export model to tflite format
kimm.export.export_tflite(reparameterized_model, 224, "model.tflite")

# Export model to onnx format (note: must be "channels_first" format)
# kimm.export.export_onnx(reparameterized_model, 224, "model.onnx")

# Enable feature extraction by setting `feature_extractor=True`
# `feature_keys` can be optionally specified
model = kimm.models.MobileOneS0(
    include_top=False, feature_extractor=True, feature_keys=["BLOCK2_S16", "BLOCK3_S32"]
)
features = model.predict(x)
for feature_name, feature in features.items():
    print(feature_name, feature.shape)  # BLOCK2_S16 (1, 14, 14, 256) ...

```

### An end-to-end example: fine-tuning an image classification model on a cats vs. dogs dataset

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1IbqfqG2NKEOKvBOznIPT1kjOdVPfThmd?usp=sharing)

Using `kimm.models.EfficientNetLiteB0`:

<div align="center">
<img width="90%" src="https://github.com/james77777778/kimm/assets/20734616/2eac0831-75bb-4790-a3af-412c3e09cf8f" alt="kimm_prediction_1">
</div>

Reference: [Transfer learning & fine-tuning (keras.io)](https://keras.io/guides/transfer_learning/#an-endtoend-example-finetuning-an-image-classification-model-on-a-cats-vs-dogs-dataset)

### Grad-CAM

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1h25VmsYDOLL6BNbRPEVOh1arIgcEoHu6?usp=sharing)

Using `kimm.models.MobileViTS`:

<div align="center">
<img width="50%" src="https://github.com/james77777778/kimm/assets/20734616/cb5022a3-aaea-4324-a9cd-3d2e63a0a6b2" alt="grad_cam">
</div>

Reference: [Grad-CAM class activation visualization (keras.io)](https://keras.io/examples/vision/grad_cam/)

## Model Zoo

|Model|Paper|Weights are ported from|API|
|-|-|-|-|
|ConvMixer|[ICLR 2022 Submission](https://arxiv.org/abs/2201.09792)|`timm`|`kimm.models.ConvMixer*`|
|ConvNeXt|[CVPR 2022](https://arxiv.org/abs/2201.03545)|`timm`|`kimm.models.ConvNeXt*`|
|DenseNet|[CVPR 2017](https://arxiv.org/abs/1608.06993)|`timm`|`kimm.models.DenseNet*`|
|EfficientNet|[ICML 2019](https://arxiv.org/abs/1905.11946)|`timm`|`kimm.models.EfficientNet*`|
|EfficientNetLite|[ICML 2019](https://arxiv.org/abs/1905.11946)|`timm`|`kimm.models.EfficientNetLite*`|
|EfficientNetV2|[ICML 2021](https://arxiv.org/abs/2104.00298)|`timm`|`kimm.models.EfficientNetV2*`|
|GhostNet|[CVPR 2020](https://arxiv.org/abs/1911.11907)|`timm`|`kimm.models.GhostNet*`|
|GhostNetV2|[NeurIPS 2022](https://arxiv.org/abs/2211.12905)|`timm`|`kimm.models.GhostNetV2*`|
|HGNet||`timm`|`kimm.models.HGNet*`|
|HGNetV2||`timm`|`kimm.models.HGNetV2*`|
|InceptionNeXt|[arXiv 2023](https://arxiv.org/abs/2303.16900)|`timm`|`kimm.models.InceptionNeXt*`|
|InceptionV3|[CVPR 2016](https://arxiv.org/abs/1512.00567)|`timm`|`kimm.models.InceptionV3`|
|LCNet|[arXiv 2021](https://arxiv.org/abs/2109.15099)|`timm`|`kimm.models.LCNet*`|
|MobileNetV2|[CVPR 2018](https://arxiv.org/abs/1801.04381)|`timm`|`kimm.models.MobileNetV2*`|
|MobileNetV3|[ICCV 2019](https://arxiv.org/abs/1905.02244)|`timm`|`kimm.models.MobileNetV3*`|
|MobileOne|[CVPR 2023](https://arxiv.org/abs/2206.04040)|`timm`|`kimm.models.MobileOne*`|
|MobileViT|[ICLR 2022](https://arxiv.org/abs/2110.02178)|`timm`|`kimm.models.MobileViT*`|
|MobileViTV2|[arXiv 2022](https://arxiv.org/abs/2206.02680)|`timm`|`kimm.models.MobileViTV2*`|
|RegNet|[CVPR 2020](https://arxiv.org/abs/2003.13678)|`timm`|`kimm.models.RegNet*`|
|RepVGG|[CVPR 2021](https://arxiv.org/abs/2101.03697)|`timm`|`kimm.models.RepVGG*`|
|ResNet|[CVPR 2015](https://arxiv.org/abs/1512.03385)|`timm`|`kimm.models.ResNet*`|
|TinyNet|[NeurIPS 2020](https://arxiv.org/abs/2010.14819)|`timm`|`kimm.models.TinyNet*`|
|VGG|[ICLR 2015](https://arxiv.org/abs/1409.1556)|`timm`|`kimm.models.VGG*`|
|ViT|[ICLR 2021](https://arxiv.org/abs/2010.11929)|`timm`|`kimm.models.VisionTransformer*`|
|Xception|[CVPR 2017](https://arxiv.org/abs/1610.02357)|`keras`|`kimm.models.Xception`|

**Comment by ibinti** [comment]
Thanks for your great work.
I am very eager to try out MobileViTV2 with ported timm imagenet weights,
Following MobileOneS0 example, I did
```
kimm.list_models("mobilevit2", weights="imagenet") 
=> ['MobileViTV2W050', 'MobileViTV2W075', 'MobileViTV2W100', 'MobileViTV2W125', 'MobileViTV2W150', 'MobileViTV2W175', 'MobileViTV2W200']
```
```
x = keras.random.uniform([1, 224, 224, 3])
model = kimm.models.MobileViTV2W100()
y = model.predict(x)
print(y.shape)
```
However, model.predict(x) throws
```
TypeError: Exception encountered when calling Reshape.call().
cannot reshape array of shape (1, 28, 28, 128) (size 100352) into shape [-1, 16, 2, 16, 2, 128] because the product of specified axis sizes (131072) does not evenly divide 100352
Arguments received by Reshape.call():
  • x=jnp.ndarray(shape=(1, 28, 28, 128), dtype=float32)
```
Thanks

**Reply by ibinti** [reply]
Well, I found the reason. MobileViTV2 expects image shape (255,255,3)
```x = keras.random.uniform([1, 256, 256, 3])```
Thanks

**Reply by ibinti** [reply]
I wonder how to handle image resizing and scaling. Is there a way to retrieve what the model expects like shape, pixel value range [0,255] float or int , or [0,1] or [-1,1]?

**Reply by james77777778** [reply]
Hey @ibinti 
Thanks for trying out this project.

Actually, there was an update a few days ago to support arbitrary sizes for MobileViT and other ViT models.
https://github.com/james77777778/keras-image-models/releases/tag/0.2.1

> I wonder how to handle image resizing and scaling. Is there a way to retrieve what the model expects like shape, pixel value range [0,255] float or int , or [0,1] or [-1,1]?

You can get the default size by:
```python
model = kimm.models.MobileNetV2W050()
print(model._default_size)  # prints 224
```

You can specify an arbitrary size for the model:
```python
# `input_shape` manner
x = keras.random.uniform([1, 256, 256, 3])
model = kimm.models.MobileNetV2W050(input_shape=[256, 256, 3])
y = model.predict(x)

# `input_tensor` manner
x = keras.random.uniform([1, 256, 256, 3])
inputs = keras.Input([256, 256, 3])
model = kimm.models.MobileNetV2W050(input_tensor=inputs)
y = model.predict(x)
```

All `kimm` models expect the value range of `[0, 255]`. The details of preprocessing can only be inspected in the source code.

- https://github.com/james77777778/keras-image-models/blob/main/kimm/_src/models/mobilevit.py#L571
- https://github.com/james77777778/keras-image-models/blob/main/kimm/_src/models/base_model.py#L147

Maybe I can add this information to the models in the future. Thanks for pointint that out!

**Reply by ibinti** [reply]
Great. Now I am ready to train kimm.models.MobileViTV2W100 and compare its performance with keras.applications.EfficientNetV2.
Thanks again!

**Comment by BWBrook** [comment]
Hi James. I've been using pretrained kimm models successfully in Docker containers. After training a model via fine-tuning and saving it to disk as a .keras file in one Docker container, if I load it up into a different Docker container to be used for inference, it triggers an automatic download of the pretrained weights again, even though they are not required for inference and the model frozen prior to saving it. 

Is there a way to avoid this behaviour? This doesn't happen if I use the keras.applications API for the same pretrained model. But I want to use kimm because it has a larger variety of available models!

**Reply by james77777778** [reply]
Hi @BWBrook 
It’s great to hear about your success with using `kimm`. I hope you're enjoying it!

Regarding your question, the default behavior is to attempt to load ImageNet weights, but this is configurable. You can set `weights=None` to skip loading, or you can use a `pathlib.Path` instance to directly load pretrained weights for your project.

You can check the implementation here to see how `kimm` handles `weights` parameter.

**Reply by BWBrook** [reply]
Thanks for answering james. Yes, I think `kimm` is great. But I didn't mean loading weights when building a model. For instance, when I'm training a new model I can use something like:

`model = kimm.models.EfficientNetV2S(include_top=False) `
This gives me pretrained imagenet weights (which I want) and I can then fine-tune on this, and then save the fine tuned model it with:
`saving.save_model(model, model_fine_tuned.keras, include_optimizer=False)`

That all works fine and saves a .keras model with the fine-tuned weights included.

But if I then try to load model_fine_tuned.keras in a new program (e.g., new Docker image) using this:
`model = saving.load_model("/code/model_fine_tuned.keras", compile=False)`
Then the model loads fine, and can be used for inference as expected BUT the load_model call triggers a download of the imagenet weights again, even though it doesn't need to. If I'd used a `keras.applications.EfficientNetV2S` instead, this download does not happen.

So I was wondering how to avoid kimm models triggering the re-downloading of the weights when they are loaded using load_model. Do you see what I'm saying? :)

**Reply by james77777778** [reply]
Oh I see. I have verified that there is a redownloading issue caused by the config json.
A simple fix is on the way and I will push a new version soon.

https://github.com/james77777778/keras-image-models/pull/54

**Reply by james77777778** [reply]
@BWBrook 
I believe `kimm==2.5.0` should fix the problem. It should be available on pypi now.

**Reply by BWBrook** [reply]
Thanks @james77777778 that's excellent. I can confirm that this has indeed fixed the re-loading problem! I really appreciate your prompt intervention. In case you're interested, this is where I'm deploying `kimm`: [MEWC - Mega Efficient Wildlife Classifier](https://github.com/zaandahl/mewc)

**Reply by james77777778** [reply]
Wow! It’s great to hear that `kimm` can be integrated into such an amazing project.

Thank you for sharing.
